src/online/unigrid.jl

```diff
@@ -23,23 +23,75 @@ function inducingpoints(
     Z = map(bounds) do lims
         LinRange(lims..., alg.m)
     end
-    return Z
+    proditer = Iterators.product(Z...)
+    return UniformGrid(proditer)
 end
 
 function updateZ!(
     ::AbstractRNG, Z::AbstractVector, alg::UniGrid, X::AbstractVector; kwargs...
 )
-    ndim = length(Z)
+    ndim = length(X[1])
     new_bounds = [extrema(x -> getindex(x, i), X) for i in 1:ndim]
-    map!(Z, Z, new_bounds) do Z_d, new_b
-        x_start = min(Z_d.start, new_b[1]) # Find the new limits
-        x_stop = max(Z_d.stop, new_b[2])
-        return LinRange(x_start, x_stop, alg.m) # readapt bounds
+    Zn = map(Z.proditer.iterators, new_bounds) do old_Z, new_b
+        x_start = min(old_Z.start, new_b[1])
+        x_stop = max(old_Z.stop, new_b[2])
+        return LinRange(x_start, x_stop, alg.m)
     end
+    Z.proditer = Iterators.product(Zn...)
     return Z
 end
 
-function updateZ(rng::AbstractRNG, Z::AbstractVector, alg::UniGrid, X::AbstractVector)
+function updateZ(
+    rng::AbstractRNG, Z::AbstractVector, alg::UniGrid, X::AbstractVector; kwargs...
+)
     Zn = deepcopy(Z)
-    return updateZ!(rng, Zn, alg, X)
+    return updateZ!(rng, Zn, alg, X; kwargs...)
+end
+
+export UniformGrid
+mutable struct UniformGrid{N,T} <: AbstractVector{T}
+    proditer::Iterators.ProductIterator{NTuple{N,LinRange{T,Int64}}}
+
+    function UniformGrid(
+        proditer::Iterators.ProductIterator{NTuple{N,LinRange{T,Int64}}}
+    ) where {N,T}
+        return new{N,T}(proditer)
+    end
+end
+
+import Base: getindex, broadcastable, eachindex, length, size, enumerate, eltype
+Base.getindex(ug::UniformGrid, i) = getelement(first(Iterators.drop(ug.proditer, i - 1)))
+getelement(x::NTuple{1,<:Real}) = only(x)
+getelement(x::NTuple) = collect(x)
+
+Base.broadcastable(ug::UniformGrid) = Base.broadcastable(ug.proditer)[:]
+
+Base.eachindex(ug::UniformGrid) = Base.OneTo(length(ug))
+
+Base.length(ug::UniformGrid) = prod(length, ug.proditer.iterators)
+# alternative: (typeof(t).parameters[1], prod(length.(ug.proditer.iterators)))
+Base.size(ug::UniformGrid) = (prod(length, ug.proditer.iterators),)
+
+Base.enumerate(ug::UniformGrid) = Base.enumerate(ug.proditer)
+
+Base.eltype(ug::UniformGrid) = typeof(ug[1])
+
+import KernelFunctions: pairwise, pairwise!
+function pairwise(d::PreMetric, x::UniformGrid)
+    return KernelFunctions.Distances_pairwise(d, x)
+end
+
+function pairwise!(out::AbstractMatrix, d::PreMetric, x::UniformGrid)
+    return KernelFunctions.Distances.pairwise!(out, d, x)
+end
```

Comment:
The `KmeansAlg` bug seems to be coming from this.

Reply:
Yeah that's exactly the same bug as https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/issues/319

Reply:
Ah I see, I found the same Julia Base issue. I am wondering why adding those methods caused a problem though, as the docs built before then.

Reply:
Actually, I suspect that this is connected to [KernelFunctions.jl#360](https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/pull/360), which I might be able to get to at some point.

Reply:
Generally I don't think these lines are needed actually, since it's an AbstractVector, this should be dealt with automatically. I think that the `T` in `UniformGrid{T}` was just wrong.

Reply:
Indeed, it was the wrong `T`. Those lines are no longer needed.